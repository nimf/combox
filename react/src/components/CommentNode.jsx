import React from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';
import { Comment, Popup, Label } from 'semantic-ui-react';

const propTypes = {
  id: PropTypes.number.isRequired,
  authorName: PropTypes.string.isRequired,
  avatarHash: PropTypes.string,
  isGuest: PropTypes.bool.isRequired,
  createdAt: PropTypes.string.isRequired,
  message: PropTypes.string.isRequired,
  children: PropTypes.node,
};
const defaultProps = {
  avatarHash: '00000000000000000000000000000000',
  children: null,
};

function CommentNode({
  id, authorName, avatarHash, isGuest, createdAt, message, children,
}) {
  return (
    <Comment>
      <Comment.Avatar
        as="a"
        src={`https://www.gravatar.com/avatar/${avatarHash}?s=100`}
      />
      <Comment.Content>
        <Comment.Author as="span">
          <span>{authorName}</span>
          { isGuest && <Label size="mini">Guest</Label>}
        </Comment.Author>
        <Comment.Metadata>
          <Popup
            trigger={<div>{moment(createdAt).fromNow()}</div>}
            content={moment(createdAt).format('lll')}
            size="mini"
            inverted
          />
        </Comment.Metadata>
        <Comment.Text>
          <p>{message}</p>
        </Comment.Text>
        <Comment.Actions>
          Reply
        </Comment.Actions>
      </Comment.Content>
      {children}
    </Comment>
  );
}

CommentNode.propTypes = propTypes;
CommentNode.defaultProps = defaultProps;

export default CommentNode;
