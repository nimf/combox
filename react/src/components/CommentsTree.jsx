import React from 'react';
import PropTypes from 'prop-types';
import { Comment } from 'semantic-ui-react';
import CommentNode from './CommentNode';

const propTypes = {
  comments: PropTypes.object.isRequired,
  currentCommentId: PropTypes.number,
};
const defaultProps = {
  currentCommentId: null,
};

function CommentsTree({
  comments, currentCommentId,
}) {
  const currentComments = comments.allIds.map(id => comments.byId[id])
    .filter(c => c.parentId === currentCommentId);
  if (currentComments.length === 0) return null;
  return (
    <Comment.Group threaded={currentCommentId === null}>
      {currentComments.map(comment =>
        (
          <CommentNode
            key={comment.id}
            id={comment.id}
            authorName={comment.authorName}
            isGuest={comment.isGuest}
            createdAt={comment.createdAt}
            message={comment.message}
          >
            <CommentsTree comments={comments} currentCommentId={comment.id} />
          </CommentNode>
        ))
      }
    </Comment.Group>
  );
}

CommentsTree.propTypes = propTypes;
CommentsTree.defaultProps = defaultProps;

export default CommentsTree;
