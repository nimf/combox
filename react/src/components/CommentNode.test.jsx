import React from 'react';
import renderer from 'react-test-renderer';
import CommentNode from './CommentNode';
import CommentsTree from './CommentsTree';
import commentsFixture from '../fixtures/comments';

const commentFixture = commentsFixture.byId[3];
const commentWithChild = commentsFixture.byId[1];

const renderComment = comment =>
  renderer.create(<CommentNode
    id={comment.id}
    authorName={comment.authorName}
    isGuest={comment.isGuest}
    createdAt={comment.createdAt}
    message={comment.message}
  >
    <CommentsTree
      comments={commentsFixture}
      currentCommentId={comment.id}
      onFocusedRendered={() => {}}
    />
  </CommentNode>).toJSON();

describe('CommentNode renders', () => {
  test('renders comment without children', () => {
    const tree = renderComment(commentFixture);
    expect(tree).toMatchSnapshot();
  });

  test('renders comment with children', () => {
    const tree = renderComment(commentWithChild);
    expect(tree).toMatchSnapshot();
  });
});
