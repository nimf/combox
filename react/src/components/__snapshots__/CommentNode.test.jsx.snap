// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentNode renders renders comment with children 1`] = `
<div
  className="comment"
>
  <a
    className="avatar"
  >
    <img
      src="https://www.gravatar.com/avatar/00000000000000000000000000000000?s=100"
    />
  </a>
  <div
    className="content"
  >
    <span
      className="author"
    >
      <span>
        Anonymous
      </span>
      <div
        className="ui mini label"
        onClick={[Function]}
      >
        Guest
      </div>
    </span>
    <div
      className="metadata"
    >
      <div
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        3 days ago
      </div>
    </div>
    <div
      className="text"
    >
      <p>
        Test comment 1
      </p>
    </div>
    <div
      className="actions"
    >
      Reply
    </div>
  </div>
  <div
    className="ui comments"
  >
    <div
      className="comment"
    >
      <a
        className="avatar"
      >
        <img
          src="https://www.gravatar.com/avatar/00000000000000000000000000000000?s=100"
        />
      </a>
      <div
        className="content"
      >
        <span
          className="author"
        >
          <span>
            Anonymous
          </span>
          <div
            className="ui mini label"
            onClick={[Function]}
          >
            Guest
          </div>
        </span>
        <div
          className="metadata"
        >
          <div
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            2 days ago
          </div>
        </div>
        <div
          className="text"
        >
          <p>
            Test comment 2
          </p>
        </div>
        <div
          className="actions"
        >
          Reply
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`CommentNode renders renders comment without children 1`] = `
<div
  className="comment"
>
  <a
    className="avatar"
  >
    <img
      src="https://www.gravatar.com/avatar/00000000000000000000000000000000?s=100"
    />
  </a>
  <div
    className="content"
  >
    <span
      className="author"
    >
      <span>
        Anonymous
      </span>
      <div
        className="ui mini label"
        onClick={[Function]}
      >
        Guest
      </div>
    </span>
    <div
      className="metadata"
    >
      <div
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        a day ago
      </div>
    </div>
    <div
      className="text"
    >
      <p>
        Test comment 3
      </p>
    </div>
    <div
      className="actions"
    >
      Reply
    </div>
  </div>
</div>
`;
